File: src/test/java/com/vjiki/music/service/FollowerServiceTest.java
package com.vjiki.music.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.List;
import java.util.UUID;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.vjiki.music.dto.FollowerResponse;
import com.vjiki.music.entity.AccessLevel;
import com.vjiki.music.entity.AuthProvider;
import com.vjiki.music.entity.User;
import com.vjiki.music.entity.UserFollow;
import com.vjiki.music.repository.UserFollowRepository;

@ExtendWith(MockitoExtension.class)
class FollowerServiceTest {

    @Mock
    private UserFollowRepository userFollowRepository;

    @InjectMocks
    private FollowerServiceImpl followerService;

    private User testUser(String email, String nickname) {
        return User.builder()
                .id(UUID.randomUUID())
                .email(email)
                .nickname(nickname)
                .accessLevel(AccessLevel.USER)
                .provider(AuthProvider.LOCAL)
                .isActive(true)
                .isVerified(false)
                .createdBy("system")
                .modifiedBy("system")
                .build();
    }

    private UserFollow follow(User follower, User followed) {
        return UserFollow.builder()
                .followerId(follower.getId())
                .followedId(followed.getId())
                .follower(follower)
                .followed(followed)
                .build();
    }

    @Nested
    @DisplayName("getFollowersByUserId")
    class GetFollowersByUserId {

        @Test
        @DisplayName("should return followers for user")
        void shouldReturnFollowersForUser() {
            UUID userId = UUID.randomUUID();
            User follower1 = testUser("follower1@example.com", "follower1");
            User follower2 = testUser("follower2@example.com", "follower2");
            User followed = testUser("followed@example.com", "followed");

            when(userFollowRepository.findByFollowedId(userId))
                    .thenReturn(List.of(follow(follower1, followed), follow(follower2, followed)));

            List<FollowerResponse> result = followerService.getFollowersByUserId(userId);

            assertThat(result).hasSize(2);
            assertThat(result)
                    .extracting(FollowerResponse::getFollowerNickname)
                    .containsExactly("follower1", "follower2");
            assertThat(result.getFirst().getFollowerId()).isEqualTo(follower1.getId());
            assertThat(result.getFirst().getFollowerEmail()).isEqualTo("follower1@example.com");
            verify(userFollowRepository).findByFollowedId(userId);
        }

        @Test
        @DisplayName("should return empty list when no followers")
        void shouldReturnEmptyListWhenNoFollowers() {
            UUID userId = UUID.randomUUID();
            when(userFollowRepository.findByFollowedId(userId)).thenReturn(List.of());

            List<FollowerResponse> result = followerService.getFollowersByUserId(userId);

            assertThat(result).isEmpty();
            verify(userFollowRepository).findByFollowedId(userId);
        }
    }
}

File: src/test/kotlin/com/vjiki/music/service/FollowerServiceTest.kt
package com.vjiki.music.service

import com.vjiki.music.entity.AccessLevel
import com.vjiki.music.entity.AuthProvider
import com.vjiki.music.entity.User
import com.vjiki.music.entity.UserFollow
import com.vjiki.music.repository.UserFollowRepository
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import java.util.*

class FollowerServiceTest : DescribeSpec({

    val userFollowRepository = mockk<UserFollowRepository>()
    val followerService = FollowerServiceImpl(userFollowRepository)

    fun createTestUser(email: String, nickname: String): User {
        return User(
            email = email,
            nickname = nickname,
            accessLevel = AccessLevel.USER,
            provider = AuthProvider.LOCAL,
            isActive = true,
            isVerified = false,
            createdBy = "system",
            modifiedBy = "system"
        )
    }

    describe("getFollowersByUserId") {
        it("should return followers for user") {
            val userId = UUID.randomUUID()
            val follower1 = createTestUser("follower1@example.com", "follower1")
            val follower2 = createTestUser("follower2@example.com", "follower2")

            val followed = createTestUser("followed@example.com", "followed")
            val userFollow1 = UserFollow(
                followerId = follower1.id,
                followedId = followed.id
            )
            val userFollow2 = UserFollow(
                followerId = follower2.id,
                followedId = followed.id
            )

            every { userFollowRepository.findByFollowedId(userId) } returns listOf(userFollow1, userFollow2)

            val result = followerService.getFollowersByUserId(userId)

            result.size shouldBe 2
            verify { userFollowRepository.findByFollowedId(userId) }
        }

        it("should return empty list when no followers") {
            val userId = UUID.randomUUID()

            every { userFollowRepository.findByFollowedId(userId) } returns emptyList()

            val result = followerService.getFollowersByUserId(userId)

            result shouldBe emptyList()
        }
    }
})

